sgkit/distance/metrics.py

```diff
@@ -0,0 +1,77 @@
+"""This module implements various distance metrics."""
+
+import numpy as np
+from numba import guvectorize
+
+from sgkit.typing import ArrayLike
+
+
+@guvectorize(["void(float64[:], float64[:], float64[:])"], "(n),(n)->()", fastmath=True)  # type: ignore
+def correlation(x: ArrayLike, y: ArrayLike, out: ArrayLike) -> None:
+    """
+    Parameters
+    ----------
+    x
+        [array-like, shape: (M,)]
+        A vector
+    y
+        [array-like, shape: (M,)]
+        Another vector
+    out
+        The output array, which has the output of pearson correlation.
+
+    Returns
+    -------
+    A scalar representing the pearson correlation coefficient between two vectors x and y.
+
+    Examples
+    --------
+    >>> from sgkit.distance.metrics import correlation
+    >>> import dask.array as da
+    >>> import numpy as np
+    >>> x = da.array([4, 3, 2, 3], dtype='i1')
+    >>> y = da.array([5, 6, 7, 0], dtype='i1')
+    >>> correlation(x, y).compute()
+    -0.2626128657194451
+
+    >>> correlation(x, x).compute()
+    1.0
+    """
+    cov = ((x - x.mean()) * (y - y.mean())).sum()
+    out[0] = cov / (x.std() * y.std()) / x.shape[0]
+
+
+@guvectorize(["void(float64[:], float64[:], float64[:])"], "(n),(n)->()", fastmath=True)  # type: ignore
+def euclidean(x: ArrayLike, y: ArrayLike, out: ArrayLike) -> None:
+    """
+
+    Parameters
+    ----------
+    x
+        [array-like, shape: (M,)]
+        A vector
+    y
+        [array-like, shape: (M,)]
+        Another vector
+    out
+        The output scalar, which has the output of euclidean between two vectors.
+
+    Returns
+    -------
+    A scalar representing the euclidean distance between two vectors x and y.
+
+    Examples
+    --------
+    >>> from sgkit.distance.metrics import euclidean
+    >>> import dask.array as da
+    >>> import numpy as np
+    >>> x = da.array([4, 3, 2, 3], dtype='i1')
+    >>> y = da.array([5, 6, 7, 0], dtype='i1')
+    >>> euclidean(x, y).compute()
+    6.6332495807108
+
+    >>> euclidean(x, x).compute()
+    0.0
+
+    """
+    out[0] = np.sqrt(((x - y) ** 2).sum())
```

Comment:
How do missing values get handled here?

Reply:
If a row has a missing value in it, it will return `nan` for all the distances, that includes that row. This behaviour is consistent with `scipy`'s `pdist`:

```
In [4]: x = np.arange(16, dtype=float).reshape((4, 4))

In [5]: x[0][1] = np.nan

In [6]: x
Out[6]:
array([[ 0., nan,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.],
       [12., 13., 14., 15.]])


In [7]: from scipy.spatial.distance import pdist, squareform

In [8]: squareform(pdist(x))
Out[8]:
array([[ 0., nan, nan, nan],
       [nan,  0.,  8., 16.],
       [nan,  8.,  0.,  8.],
       [nan, 16.,  8.,  0.]])

In [9]: from sgkit.distance.api import pdist

In [10]: pdist(x)
Out[10]:
array([[nan, nan, nan, nan],
       [nan,  0.,  8., 16.],
       [nan,  8.,  0.,  8.],
       [nan, 16.,  8.,  0.]])

```

Reply:
Hm I see.  That definitely won't be the behavior most users would want so the two ways around this I see are:

1. Force them to use some kind of imputation function first
2. Implement nan-aware metrics

I had been assuming 2 would make the most sense where missing values in either vector are ignored rather than leading to nan propagation.  Option 1 certainly seems to be a more popular choice in other domains but the missingness in genetics is typically low enough that nan-aware semantics become appealing too.  I could go either way on it though committing to operating on integers makes the nan-aware approach more appealing.  I had envisioned our implementations looking like this until having a good reason to change course: https://github.com/cggh/scikit-allel/blob/961254bd583572eed7f9bd01060e53a8648e620c/allel/opt/stats.pyx#L58 (how scikit-allel ignores missing values in pairwise LD calcs).

Reply:
I have incorporated this in the latest commit.